main.tf

```diff
@@ -35,21 +35,43 @@ resource "azurerm_network_security_rule" "predefined_rules" {
   source_port_ranges                         = lookup(var.predefined_rules[count.index], "source_port_range", "*") == "*" ? null : split(",", var.predefined_rules[count.index].source_port_range)
 }
 
+resource "azurerm_network_security_rule" "predefined_rules_for" {
+  for_each = { for value in var.predefined_rules : value.name => value if var.use_for_each }
+
+  access                                     = element(var.rules[lookup(each.value, "name")], 1)
+  direction                                  = element(var.rules[lookup(each.value, "name")], 0)
+  name                                       = each.value.name
+  network_security_group_name                = azurerm_network_security_group.nsg.name
+  priority                                   = each.value.priority
```

Comment:
I understand the reason why the `priority` in for_each version is different than it's counterpart in count version, because we're lacking of the index so we cannot calculate the priority when it's absent. We might need to warn our users the `priority` is required if you choose `var.for_each = true`.